go/worker/storage/committee/checkpoint_sync.go

```diff
@@ -0,0 +1,268 @@
+package committee
+
+import (
+	"container/heap"
+	"context"
+	"errors"
+	"io"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/oasisprotocol/oasis-core/go/runtime/committee"
+	schedulerApi "github.com/oasisprotocol/oasis-core/go/scheduler/api"
+	storageApi "github.com/oasisprotocol/oasis-core/go/storage/api"
+	storageClient "github.com/oasisprotocol/oasis-core/go/storage/client"
+	"github.com/oasisprotocol/oasis-core/go/storage/mkvs/checkpoint"
+)
+
+const (
+	checkpointStatusDone = 0
+	checkpointStatusNext = 1
+	checkpointStatusBail = 2
+
+	// LogEventCheckpointSyncSuccess is a log event value that signals that checkpoint sync was successful.
+	LogEventCheckpointSyncSuccess = "worker/storage/checkpoint-sync-success"
+)
+
+var (
+	// ErrNoUsableCheckpoints is the error returned when none of the checkpoints could be synced.
+	ErrNoUsableCheckpoints = errors.New("storage: no checkpoint could be synced")
+)
+
+type restoreResult struct {
+	done bool
+	err error
+}
+
+type chunkHeap struct {
+	array []*checkpoint.ChunkMetadata
+	length int
+}
+
+func (h chunkHeap) Len() int { return h.length }
+func (h chunkHeap) Less(i, j int) bool { return h.array[i].Index < h.array[j].Index }
+func (h chunkHeap) Swap(i, j int) { h.array[i], h.array[j] = h.array[j], h.array[i] }
+
+func (h *chunkHeap) Push(x interface{}) {
+	h.array[h.length] = x.(*checkpoint.ChunkMetadata)
+	h.length++
+}
+
+func (h *chunkHeap) Pop() interface{} {
+	ret := h.array[h.length-1]
+	h.array[h.length] = nil
+	h.length--
+	return ret
+}
+
+func (n *Node) nodeWorker(
+	ctx context.Context,
+	clientConn *committee.ClientConnWithMeta,
+	chunkDispatchCh chan *checkpoint.ChunkMetadata,
+	chunkReturnCh chan *checkpoint.ChunkMetadata,
+	errorCh chan int,
+) {
+	conn := storageApi.NewStorageClient(clientConn.ClientConn)
+	for {
+		var chunk *checkpoint.ChunkMetadata
+		var ok bool
+		select {
+		case <-ctx.Done():
+			return
+		case chunk, ok = <-chunkDispatchCh:
+			if !ok {
+				return
+			}
+		}
+
```

Comment:
I think currently `chunk` can be nil here (see my other comment), but this is not handled?